File: Qash.Mobile/qash/lib/features/transactions/domain/entities/transaction.dart
enum TransactionType { income, expense, transfer }

class TransactionEntity {
  final String id;
  final String walletId;
  final String walletName;
  final String? toWalletId;
  final String? toWalletName;
  final String userId;
  final double amount;
  final double? toAmount;
  final String walletCurrency;
  final String? toWalletCurrency;
  final TransactionType type;
  final String categoryId;
  final String categoryName;
  final String? description;
  final DateTime transactionDate;

  const TransactionEntity({
    required this.id,
    required this.walletId,
    required this.walletName,
    this.toWalletId,
    this.toWalletName,
    required this.userId,
    required this.amount,
    this.toAmount,
    this.walletCurrency = 'USD',
    this.toWalletCurrency,
    required this.type,
    required this.categoryId,
    required this.categoryName,
    this.description,
    required this.transactionDate,
  });

  bool get isIncome => type == TransactionType.income;
  bool get isExpense => type == TransactionType.expense;
  bool get isTransfer => type == TransactionType.transfer;

  bool get isCrossCurrencyTransfer =>
      isTransfer &&
      toWalletCurrency != null &&
      toWalletCurrency!.isNotEmpty &&
      walletCurrency.isNotEmpty &&
      toWalletCurrency!.toUpperCase() != walletCurrency.toUpperCase();

  double get creditAmount => toAmount ?? amount;
}

File: Qash.API/Domain/Entities/Transaction.cs
using Qash.API.Domain.Common;
using Qash.API.Domain.Enums;

namespace Qash.API.Domain.Entities;

public class Transaction : BaseEntity
{
    public Guid WalletId { get; set; }

    public Wallet Wallet { get; set; } = null!;

    public Guid? ToWalletId { get; set; }

    public Wallet? ToWallet { get; set; }

    /// <summary>Links paired transfer legs (outgoing expense + incoming income).</summary>
    public Guid? TransferGroupId { get; set; }

    /// <summary>Points to the counterpart transaction in a transfer pair.</summary>
    public Guid? LinkedTransactionId { get; set; }

    public Guid ApplicationUserId { get; set; }

    public ApplicationUser ApplicationUser { get; set; } = null!;

    public CategoryType TransactionType { get; set; }

    public decimal Amount { get; set; }

    /// <summary>
    /// Amount credited to <see cref="ToWallet"/> for transfers (destination currency).
    /// </summary>
    public decimal? ToAmount { get; set; }

    /// <summary>Original transaction currency (source wallet).</summary>
    public string SourceCurrency { get; set; } = "USD";

    /// <summary>Destination wallet currency for transfers.</summary>
    public string? DestinationCurrency { get; set; }

    /// <summary>Amount normalized to the base currency (USD) at transaction time.</summary>
    public decimal AmountInBaseCurrency { get; set; }

    /// <summary>
    /// Exchange rate applied: destination per source unit for transfers,
    /// or source-to-base rate for income/expense.
    /// </summary>
    public decimal? ExchangeRateUsed { get; set; }

    public Guid CategoryId { get; set; }

    public Category Category { get; set; } = null!;

    public string Description { get; set; } = string.Empty;

    public DateTime TransactionDate { get; set; } = DateTime.UtcNow;
}
